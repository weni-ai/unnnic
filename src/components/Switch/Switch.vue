<template>
  <div class="unnnic-switch-wrapper">
    <label>
      <input
        class="unnnic-switch"
        type="checkbox"
        :disabled="disabled"
        :checked="modelValue"
        @change="toggleState"
        v-bind="pick($attrs, ['id', 'name'])"
      />

      <p
        v-if="label || textLeft || textRight"
        :class="[
          'unnnic-switch__label',
          { 'unnnic-switch__label--disabled': disabled },
        ]"
        data-testid="switch-label"
      >
        {{ label }}
        {{ textLeft }}
        {{ textRight }}
      </p>
    </label>

    <p
      v-if="helper"
      class="unnnic-switch__helper"
    >
      {{ helper }}
    </p>
  </div>
</template>

<script>
import { pick } from 'lodash';

export default {
  name: 'UnnnicSwitch',
  props: {
    size: {
      type: String,
      default: 'medium',
      validator(value) {
        return ['small', 'medium'].indexOf(value) !== -1;
      },
    },

    label: {
      type: String,
      default: '',
    },

    helper: {
      type: String,
      default: '',
    },

    textLeft: {
      type: String,
      default: '',
    },
    textRight: {
      type: String,
      default: '',
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    modelValue: {
      type: Boolean,
      default: false,
    },
    useVModel: {
      type: Boolean,
    },
  },
  emits: ['update:model-value'],
  data() {
    return {
      isActive: false,
    };
  },

  watch: {
    modelValue: {
      immediate: true,
      handler() {
        this.isActive = this.modelValue;
      },
    },
  },

  methods: {
    pick,

    toggleState() {
      if (!this.disabled) {
        if (this.useVModel) {
          this.$emit('update:model-value', !this.isActive);
        } else {
          this.isActive = !this.isActive;
          this.$emit('update:model-value', this.isActive);
        }
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;

$switch-width: 38px;
$switch-height: 21px;

.unnnic-switch-wrapper {
  display: flex;
  flex-direction: column;

  label {
    display: flex;
    align-items: center;
    column-gap: $unnnic-space-2;
  }
}

.unnnic-switch {
  appearance: none;
  width: $switch-width;
  height: $switch-height;
  margin: 0;
  background-color: $unnnic-color-bg-muted;
  border-radius: $unnnic-radius-3;
  box-sizing: border-box;
  outline: none;

  background-image: url('@/assets/icons/switch-checked.svg');
  background-repeat: no-repeat;
  background-position: 2px center;

  transition: 200ms linear background-position, 200ms linear background-color;

  &:checked {
    background-color: $unnnic-color-bg-active;
    background-position: 20px center;
  }

  &:disabled {
    background-color: $unnnic-color-bg-muted;
    background-image: url('@/assets/icons/switch-checked-disabled.svg');
  }

  &__label {
    margin: 0;
    font: $unnnic-font-body;
    color: $unnnic-color-fg-emphasized;

    &--disabled {
      color: $unnnic-color-fg-muted;
    }
  }

  &__helper {
    margin: 0;
    margin-left: $switch-width + $unnnic-space-2;
    font: $unnnic-font-caption-2;
    color: $unnnic-color-fg-base;
  }
}
</style>
