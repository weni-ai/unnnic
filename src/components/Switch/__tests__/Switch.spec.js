import { beforeEach, describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';

import Switch from '../Switch.vue';

const createWrapper = (props) => {
  return mount(Switch, { props });
};

describe('Switch', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = createWrapper({ modelValue: false });
  });
  it('should render the component correctly with default props', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.findComponent({ name: 'Icon' }).exists()).toBe(true);
    expect(wrapper.text()).toBe('');
  });

  it('should render the left text (textLeft) when provided', async () => {
    await wrapper.setProps({ textLeft: 'Left Text' });

    // This check ensures that the icon is placed before the switch text
    const switchChildren = wrapper.findComponent(Switch).element.children;
    expect(switchChildren[0].getAttribute('data-testid')).toBe(
      'switch-text-left',
    );
    expect(switchChildren[1].getAttribute('data-testid')).toBe('switch-icon');
  });

  it('should render the right text (textRight) when provided', async () => {
    await wrapper.setProps({ textRight: 'Right Text' });

    // This check ensures that the icon is placed after the switch text
    const switchChildren = wrapper.findComponent(Switch).element.children;
    expect(switchChildren[0].getAttribute('data-testid')).toBe('switch-icon');
    expect(switchChildren[1].getAttribute('data-testid')).toBe(
      'switch-text-right',
    );
  });

  it('should render the correct icon when the switch is disabled/enabled', async () => {
    await wrapper.setProps({ disabled: true });
    const disabledIconComponent = wrapper.findComponent({ name: 'Icon' });
    expect(disabledIconComponent.props('icon')).toBe('switch-default-disabled');

    await wrapper.setProps({ modelValue: true });
    const enabledIconComponent = wrapper.findComponent({ name: 'Icon' });
    expect(enabledIconComponent.props('icon')).toBe('switch-selected-disabled');
  });

  it('should render the correct icon based on the isActive state', async () => {
    expect(wrapper.vm.isActive).toBe(false);
    await wrapper.setProps({ modelValue: true });
    expect(wrapper.vm.isActive).toBe(true);
    expect(wrapper.findComponent({ name: 'Icon' }).props('icon')).toBe(
      'switch-default',
    );
  });

  it('should render the correct icon size based on the size prop', async () => {
    await wrapper.setProps({ size: 'small' });
    expect(wrapper.findComponent({ name: 'Icon' }).props('size')).toBe('sm');

    await wrapper.setProps({ size: 'medium' });
    expect(wrapper.findComponent({ name: 'Icon' }).props('size')).toBe('md');
  });

  it('should change the icon color based on the state', async () => {
    expect(wrapper.findComponent({ name: 'Icon' }).props('scheme')).toBe(
      'neutral-soft',
    );

    await wrapper.setProps({ modelValue: true });
    expect(wrapper.findComponent({ name: 'Icon' }).props('scheme')).toBe(
      'brand-weni',
    );

    await wrapper.setProps({ disabled: true });
    expect(wrapper.findComponent({ name: 'Icon' }).props('scheme')).toBe(
      'neutral-soft',
    );
  });

  it('should toggle isActive state and emit the correct event when toggleState is called', async () => {
    const initialIsActive = wrapper.vm.isActive;
    const switchIcon = wrapper.findComponent('[ data-testid="switch-icon"]');
    await switchIcon.trigger('click');
    expect(wrapper.emitted('update:model-value')).toBeTruthy();
    expect(wrapper.vm.isActive).toBe(!initialIsActive);
  });

  it('should not change state when disabled is true', async () => {
    await wrapper.setProps({ disabled: true });
    const switchIcon = wrapper.findComponent('[ data-testid="switch-icon"]');
    await switchIcon.trigger('click');
    expect(wrapper.emitted('update:model-value')).toBeFalsy();
    expect(wrapper.vm.isActive).toBe(false);
  });

  it('should update isActive state when modelValue changes', async () => {
    expect(wrapper.vm.isActive).toBe(false);

    await wrapper.setProps({ modelValue: true });
    expect(wrapper.vm.isActive).toBe(true);
  });

  it('should correctly use v-model to manage the isActive state', async () => {
    await wrapper.setProps({ useVModel: true });
    await wrapper.vm.toggleState();
    expect(wrapper.emitted('update:model-value')[0]).toEqual([true]);
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
