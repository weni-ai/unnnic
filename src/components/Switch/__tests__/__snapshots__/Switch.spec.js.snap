// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch > should match the snapshot 1`] = `
"<div data-v-e4356c9d="" class="unnnic-switch-wrapper"><label data-v-e4356c9d=""><input data-v-e4356c9d="" class="unnnic-switch" type="checkbox">
    <!--v-if-->
  </label>
  <!--v-if-->
</div>"
`;
